aspnetcore/signalr/authn-and-authz/sample/Startup.cs

```diff
@@ -0,0 +1,118 @@
+
+namespace SignalRAuthenticationSample
+{
+    public class Startup
+    {
+        // We use a key generated on this server during startup to secure our tokens.
+        // This means that if the app restarts, existing tokens become invalid. It also won't work
+        // when using multiple servers.
+        public static readonly SymmetricSecurityKey SecurityKey = new SymmetricSecurityKey(Guid.NewGuid().ToByteArray());
+
+        public Startup(IConfiguration configuration)
+        {
+            Configuration = configuration;
+        }
+
+        public IConfiguration Configuration { get; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddDbContext<ApplicationDbContext>(options =>
+                options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));
+
+            services.AddIdentity<ApplicationUser, IdentityRole>()
+                .AddEntityFrameworkStores<ApplicationDbContext>()
+                .AddDefaultTokenProviders();
+
+            services.AddAuthentication(options =>
+                {
+                    // Identity made Cookie authentication the default, we want JWT Bearer Auth to be the default.
+                    options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
+                    options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
+                })
+                .AddJwtBearer(options =>
+                {
+                    // Configure JWT Bearer Auth to expect our security key
+                    options.TokenValidationParameters =
+                        new TokenValidationParameters
+                        {
+                            LifetimeValidator = (before, expires, token, parameters) => expires > DateTime.UtcNow,
+                            ValidateAudience = false,
+                            ValidateIssuer = false,
+                            ValidateActor = false,
+                            ValidateLifetime = true,
+                            IssuerSigningKey = SecurityKey
+                        };
+
+                    // We have to hook the OnMessageReceived event in order to allow the JWT authentication handler
+                    // to read the access token from the query string when a WebSocket or Server-Sent Events request comes in.
+                    options.Events = new JwtBearerEvents
+                    {
+                        OnMessageReceived = context =>
+                        {
+                            var accessToken = context.Request.Query["access_token"];
+
+                            // If the request is for our hub...
+                            if (!string.IsNullOrEmpty(accessToken) &&
+                                (context.HttpContext.Request.Path.StartsWithSegments("/hubs/chat")))
```

Comment:
Is this what we expect people to do? Check the path?

Reply:
Our examples use `IsWebSocketRequest` but there's a problem with that. If you have the following:

```csharp
app.UseAuthentication();
app.UseSignalR();
```

Then `IsWebSocketRequest` won't be set when your event is fired because it's the WebSocket middleware that does that and SignalR adds the WebSocket middleware, so it ends up **after** authentication. This can be fixed by adding `UseWebSockets` above `UseAuthentication` but that seemed harder to explain and uglier.